From 5b0bb3102f3814debf2098ba00f737715b1acc24 Mon Sep 17 00:00:00 2001
From: Arockiasathish S <arocs@qti.qualcomm.com>
Date: Wed, 17 Jun 2026 12:08:49 +0530
Subject: [PATCH] [RDK][QMESH] handle topology notification and improve TLV
 parsing robustness

- Add support for TOPOLOGY_NOTIFICATION_MESSAGE in backhaul manager
  and topology task

- Implement topology notification handler:
  - Log incoming notification
  - Trigger TOPOLOGY_QUERY_MESSAGE to sender as per IEEE 1905.1 spec

- Allow topology discovery in RDKB builds even when controller is not discovered

- Improve TLV parsing in CmduMessageRx:
  - Skip zero-length TLVs (except End-of-Message)
  - Prevent buffer underflow issues in generated init() logic

Signed-off-by: Arockiasathish S <arocs@qti.qualcomm.com>
---
 .../backhaul_manager/backhaul_manager.cpp     |  1 +
 .../beerocks/slave/tasks/topology_task.cpp    | 27 +++++++++++++++++++
 .../src/beerocks/slave/tasks/topology_task.h  |  8 ++++++
 framework/tlvf/src/src/CmduMessageRx.cpp      | 17 +++++++++++-
 4 files changed, 52 insertions(+), 1 deletion(-)

diff --git a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
index a8fe0f0ae..e48466236 100644
--- a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
+++ b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
@@ -303,6 +303,7 @@ bool BackhaulManager::thread_init()
             ieee1905_1::eMessageType::CHANNEL_SELECTION_REQUEST_MESSAGE,
             ieee1905_1::eMessageType::HIGHER_LAYER_DATA_MESSAGE,
             ieee1905_1::eMessageType::TOPOLOGY_DISCOVERY_MESSAGE,
+            ieee1905_1::eMessageType::TOPOLOGY_NOTIFICATION_MESSAGE,
             ieee1905_1::eMessageType::TOPOLOGY_QUERY_MESSAGE,
             ieee1905_1::eMessageType::VENDOR_SPECIFIC_MESSAGE,
             ieee1905_1::eMessageType::UNASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE,
diff --git a/agent/src/beerocks/slave/tasks/topology_task.cpp b/agent/src/beerocks/slave/tasks/topology_task.cpp
index 35f5331e3..f127f7d65 100644
--- a/agent/src/beerocks/slave/tasks/topology_task.cpp
+++ b/agent/src/beerocks/slave/tasks/topology_task.cpp
@@ -113,11 +113,13 @@ void TopologyTask::work()
         send_topology_notification();
     }
 
+#ifndef BEEROCKS_RDKB
     // Don't send topology discovery if the Controller hasn't been discovered and in Controller Only Mode
     if ((db->controller_info.bridge_mac == network_utils::ZERO_MAC) &&
         (db->device_conf.management_mode != BPL_MGMT_MODE_MULTIAP_CONTROLLER)) {
         return;
     }
+#endif // BEEROCKS_RDKB
 
     // Send topology discovery every 60 seconds according to IEEE_Std_1905.1-2013 specification
     if (now - m_periodic_discovery_timestamp >
@@ -160,6 +162,10 @@ bool TopologyTask::handle_cmdu(ieee1905_1::CmduMessageRx &cmdu_rx, uint32_t ifac
         handle_topology_query(cmdu_rx, src_mac);
         break;
     }
+    case ieee1905_1::eMessageType::TOPOLOGY_NOTIFICATION_MESSAGE: {
+        handle_topology_notification(cmdu_rx, src_mac);
+        break;
+    }
     case ieee1905_1::eMessageType::VENDOR_SPECIFIC_MESSAGE: {
         return handle_vendor_specific(cmdu_rx, src_mac, beerocks_header);
     }
@@ -251,6 +257,27 @@ void TopologyTask::handle_topology_discovery(ieee1905_1::CmduMessageRx &cmdu_rx,
     }
 }
 
+void TopologyTask::handle_topology_notification(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                                const sMacAddr &src_mac)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(INFO) << "Received TOPOLOGY_NOTIFICATION_MESSAGE from " << src_mac 
+              << ", mid=" << std::hex << mid;
+    
+    // According to IEEE 1905.1 specification, when a device receives a Topology Notification,
+    // it should send a Topology Query to the sender to get updated topology information
+    auto cmdu_tx_header =
+        m_cmdu_tx.create(0, ieee1905_1::eMessageType::TOPOLOGY_QUERY_MESSAGE);
+    if (!cmdu_tx_header) {
+        LOG(ERROR) << "Failed creating topology query message";
+        return;
+    }
+    
+    auto db = AgentDB::get();
+    LOG(DEBUG) << "Sending TOPOLOGY_QUERY_MESSAGE to " << src_mac;
+    m_btl_ctx.send_cmdu_to_broker(m_cmdu_tx, src_mac, db->bridge.mac);
+}
+
 void TopologyTask::handle_topology_query(ieee1905_1::CmduMessageRx &cmdu_rx,
                                          const sMacAddr &src_mac)
 {
diff --git a/agent/src/beerocks/slave/tasks/topology_task.h b/agent/src/beerocks/slave/tasks/topology_task.h
index e5228a32c..ac8b4d158 100644
--- a/agent/src/beerocks/slave/tasks/topology_task.h
+++ b/agent/src/beerocks/slave/tasks/topology_task.h
@@ -59,6 +59,14 @@ private:
     */
     void handle_topology_query(ieee1905_1::CmduMessageRx &cmdu_rx, const sMacAddr &src_mac);
 
+    /**
+    * @brief Handles 1905 Topology Notification message.
+    * 
+    * @param[in] cmdu_rx Received CMDU.
+    * @param[in] src_mac MAC address of the message sender.
+    */
+    void handle_topology_notification(ieee1905_1::CmduMessageRx &cmdu_rx, const sMacAddr &src_mac);
+
     /**
      * @brief Handles Vendor Specific messages. 
      * 
diff --git a/framework/tlvf/src/src/CmduMessageRx.cpp b/framework/tlvf/src/src/CmduMessageRx.cpp
index 3bdf17c0c..9c78d8b30 100644
--- a/framework/tlvf/src/src/CmduMessageRx.cpp
+++ b/framework/tlvf/src/src/CmduMessageRx.cpp
@@ -649,7 +649,22 @@ std::shared_ptr<BaseClass> CmduMessageRx::parseNextTlv(wfa_map::eTlvTypeMap tlv_
 std::shared_ptr<BaseClass> CmduMessageRx::parseNextTlv()
 {
     auto tlv_type = getNextTlvType();
-
+#ifdef BEEROCKS_RDKB
+    auto tlv_length = getNextTlvLength();
+    // For zero-length TLVs (except End of Message which also has length 0),
+    // the generated init() code has a buffer underflow bug when calculating
+    // remaining length. Skip them by manually advancing the buffer pointer
+    // past the TLV header (3 bytes: 1 type + 2 length).
+    if (tlv_length == 0 &&
+        tlv_type != static_cast<int>(ieee1905_1::eTlvType::TLV_END_OF_MESSAGE)) {
+        LOG(DEBUG) << "Zero-length TLV type=" << tlv_type << ", skipping";
+        if (!msg.prevClass()->buffPtrIncrementSafe(sizeof(sTlvHeader))) {
+            LOG(ERROR) << "buffPtrIncrementSafe failed for zero-length TLV";
+            return nullptr;
+        }
+        return parseNextTlv();
+    }
+#endif
     if (ieee1905_1::eTlvTypeValidate::check(tlv_type)) {
         return parseNextTlv(ieee1905_1::eTlvType(tlv_type));
     } else if (wfa_map::eTlvTypeMapValidate::check(tlv_type)) {
-- 
2.34.1

